src/conn/conn_log.c

```diff
@@ -400,12 +401,21 @@ __log_remove_once(WT_SESSION_IMPL *session, uint32_t backup_file)
     min_lognum = backup_file == 0 ? WT_MIN(log->ckpt_lsn.l.file, log->sync_lsn.l.file) :
                                     WT_MIN(log->ckpt_lsn.l.file, backup_file);
 
+    /*
+     * Take a consistent view of the current configuration. If another thread reconfigures the
+     * connection in parallel we'll handle that change on the next call to this function.
+     */
+    __wt_spin_lock(session, &conn->reconfig_lock);
+    dbg_ckpt_cnt = conn->debug_ckpt_cnt;
+    dbg_ckpt_retain = FLD_ISSET(conn->debug_flags, WT_CONN_DEBUG_CKPT_RETAIN);
+    dbg_log_cnt = conn->debug_log_cnt;
+    __wt_spin_unlock(session, &conn->reconfig_lock);
+
     /* Adjust the number of log files to retain based on debugging options. */
-    WT_ORDERED_READ(dbg_val, conn->debug_ckpt_cnt);
-    if (FLD_ISSET(conn->debug_flags, WT_CONN_DEBUG_CKPT_RETAIN) && dbg_val != 0)
-        min_lognum = WT_MIN(conn->debug_ckpt[dbg_val - 1].l.file, min_lognum);
-    WT_ORDERED_READ(dbg_val, conn->debug_log_cnt);
-    if (dbg_val != 0) {
+    if (dbg_ckpt_retain && dbg_ckpt_cnt != 0)
+        min_lognum = WT_MIN(conn->debug_ckpt[dbg_ckpt_cnt - 1].l.file, min_lognum);
```

Comment:
This access to conn->debug_ckpt is outside of the lock but we know it's safe to access it like this (per [this comment](https://github.com/wiredtiger/wiredtiger/blob/fe22c23199fd0f6462a148c51b0a9694857a19bc/src/conn/conn_api.c#L2032) we'll never have a non-zero dbg_ckpt_cnt and an unallocated conn_debug_ckpt array).
Do we play it safe and wrap the whole section in the lock, or try and minimise the time spent holding the lock?

Reply:
Play it safe and lock the whole section, conn_reconfig should pay the price of waiting.